lib/Language/Souffle/Interpreted.hs

```diff
@@ -178,47 +191,62 @@ instance MonadSouffle SouffleM where
       souffleTempDir <- liftIO $ do
         tmpDir <- getCanonicalTemporaryDirectory
         createTempDirectory tmpDir "souffle-haskell"
-      let factDir = souffleTempDir </> "fact"
-          outDir  = souffleTempDir </> "out"
+
+      factDir <- fromMaybe (souffleTempDir </> "fact") <$> asks cfgFactDir
+      outDir <- fromMaybe (souffleTempDir </> "out") <$> asks cfgOutputDir
       liftIO $ do
         createDirectoryIfMissing True factDir
         createDirectoryIfMissing True outDir
       mSouffleBin <- asks cfgSouffleBin
       liftIO $ forM mSouffleBin $ \souffleBin ->
-        fmap Handle $ newIORef $ HandleData
-          { soufflePath = souffleBin
-          , basePath    = souffleTempDir
-          , factPath    = factDir
-          , outputPath  = outDir
-          , datalogExec = datalogExecutable
-          , noOfThreads = 1
-          }
+        Handle <$> ((,,)
+          <$> (newIORef $ HandleData
+                { soufflePath = souffleBin
+                , basePath    = souffleTempDir
+                , factPath    = factDir
+                , outputPath  = outDir
+                , datalogExec = datalogExecutable
+                , noOfThreads = 1
+                })
+          <*> newIORef Nothing
+          <*> newIORef Nothing)
   {-# INLINABLE init #-}
 
-  run (Handle ref) = liftIO $ do
-    handle <- readIORef ref
+  run (Handle (refHandleData,refHandleStdOut,refHandleStdErr)) = liftIO $ do
+    handle <- readIORef refHandleData
     -- Invoke the souffle binary using parameters, supposing that the facts
     -- are placed in the factPath, rendering the output into the outputPath.
-    callCommand $
-      printf "%s -F%s -D%s -j%d %s"
-        (soufflePath handle)
-        (factPath handle)
-        (outputPath handle)
-        (noOfThreads handle)
-        (datalogExec handle)
+    let processToRun =
+          (shell
+            (printf "%s -F%s -D%s -j%d %s"
+              (soufflePath handle)
+              (factPath handle)
+              (outputPath handle)
+              (noOfThreads handle)
+              (datalogExec handle)))
+            { std_in  = NoStream
+            , std_out = CreatePipe
+            , std_err = CreatePipe
+            }
+    (_, mStdOutHandle, mStdErrHandle, processHandle) <- createProcess_ "souffle-haskell" processToRun
+    writeIORef refHandleStdOut mStdOutHandle
+    writeIORef refHandleStdErr mStdErrHandle
```

Comment:
What if you run `run` twice? Cleanup will no longer have access to old handles? I think we will leak open file handles here.

Reply:
Changed to maintain the content of stdout and stderr.